reactive-streams/src/main/scala/fs2/interop/reactivestreams/StreamSubscriber.scala

```diff
@@ -74,85 +83,116 @@ object StreamSubscriber {
     def onFinalize: F[Unit]
 
     /** producer for downstream */
-    def dequeue1: F[Either[Throwable, Option[A]]]
+    def dequeue1: F[Chunk[Either[Throwable, Option[A]]]]
 
     /** downstream stream */
     def stream(subscribe: F[Unit])(implicit ev: ApplicativeError[F, Throwable]): Stream[F, A] =
       Stream.bracket(subscribe)(_ => onFinalize) >> Stream
-        .eval(dequeue1)
+        .evalUnChunk(dequeue1)
         .repeat
         .rethrow
         .unNoneTerminate
   }
 
-  private[reactivestreams] def fsm[F[_], A](implicit F: Concurrent[F]): F[FSM[F, A]] = {
-    type Out = Either[Throwable, Option[A]]
+  private[reactivestreams] def fsm[F[_], A](
+      q: Queue[F, Either[Throwable, Option[A]]],
+      batchSize: Int
+  )(implicit F: Concurrent[F]): F[FSM[F, A]] = {
 
     sealed trait Input
     case class OnSubscribe(s: Subscription) extends Input
     case class OnNext(a: A) extends Input
     case class OnError(e: Throwable) extends Input
     case object OnComplete extends Input
     case object OnFinalize extends Input
-    case class OnDequeue(response: Deferred[F, Out]) extends Input
+    case object OnDequeue extends Input
 
     sealed trait State
     case object Uninitialized extends State
     case class Idle(sub: Subscription) extends State
-    case class RequestBeforeSubscription(req: Deferred[F, Out]) extends State
-    case class WaitingOnUpstream(sub: Subscription, elementRequest: Deferred[F, Out]) extends State
+    case class NextBatch(sub: Subscription) extends State
+    case class Receiving(sub: Subscription) extends State
+    case object RequestBeforeSubscription extends State
+    case class WaitingOnUpstream(sub: Subscription) extends State
     case object UpstreamCompletion extends State
     case object DownstreamCancellation extends State
     case class UpstreamError(err: Throwable) extends State
 
-    def step(in: Input): State => (State, F[Unit]) =
+    def step(in: Input, counter: Long): State => (State, F[Unit]) =
       in match {
         case OnSubscribe(s) => {
-          case RequestBeforeSubscription(req) => WaitingOnUpstream(s, req) -> F.delay(s.request(1))
-          case Uninitialized                  => Idle(s) -> F.unit
+          case RequestBeforeSubscription => WaitingOnUpstream(s) -> F.delay(s.request(batchSize))
```

Comment:
This case is interesting. So if we request before we've subscribed, we end up in state `RequestBeforeSubscription`. And then when we do subscribe, we go into `WaitingOnUpstream` rather than `Idle`. I haven't fully pieced together why we need both `Idle` and `WaitingOnUpstream`.

If we're `Idle` when we hit `Input.OnDequeue`, then we switch to `WaitingOnUpstream`. Do we really need both `Idle` and `WaitingOnUpstream`? This might well be needed in order for us to be in the right state to adhere to all the reactive streams rules; I was just looking for unneeded complexity and was curious if both these states are needed.

Reply:
Good observation, I'm not sure we do need both now that we don't have the `Deferred[F, Out]` to manage. It seems like we could replace `WaitingOnUpstream` with `Receiving`.

Reply:
Yes. This is interesting. This was already the behavior in there. I wonder why that happens exactly.